src/web-extensions.c

```diff
@@ -0,0 +1,696 @@
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  we_session = (WebExtensionsSession *)session;
+
+  if (we_session->state != WEB_EXTENSIONS_SESSION_STATE_INIT)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session already started");
+      return TRUE;
+    }
+
+  we_session->state = WEB_EXTENSIONS_SESSION_STATE_STARTING;
+  g_object_set_data_full (G_OBJECT (request), "session", g_object_ref (session), g_object_unref);
+  g_object_set_data_full (G_OBJECT (request), "name", g_strdup (arg_name), g_free);
+  g_object_set_data_full (G_OBJECT (request), "extension-or-origin", g_strdup (arg_extension_or_origin), g_free);
+
+  request_export (request, g_dbus_method_invocation_get_connection (invocation));
+  xdp_dbus_web_extensions_complete_start (object, invocation, request->id);
+
+  task = g_task_new (object, NULL, NULL, NULL);
+  g_task_set_task_data (task, g_object_ref (request), g_object_unref);
+  g_task_run_in_thread (task, handle_start_in_thread);
+
+  return TRUE;
+}
+
+
+static gboolean
+handle_get_pipes (XdpDbusWebExtensions *object,
+                  GDBusMethodInvocation *invocation,
+                  GUnixFDList *fd_list,
+                  const char *arg_session_handle,
+                  GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  Session *session;
+  WebExtensionsSession *we_session;
+  g_autoptr(GUnixFDList) out_fd_list = NULL;
+
+  session = acquire_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  we_session = (WebExtensionsSession *)session;
+
+  if (we_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTED)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session not started");
+      return TRUE;
+    }
+
+  if (we_session->standard_input < 0 ||
+      we_session->standard_output < 0 ||
+      we_session->standard_error < 0)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "GetPipes already called");
+      return TRUE;
+    }
+
+  int fds[3] = {
```

Comment:
I'd rather not have body variable declarations like this. Please move the `int fds[3]` to the top of the function.